File: frontend/src/types/review-analysis.ts
import type { SiteCode } from "@/types/selection";

export type AnalysisStatus = "PENDING" | "RUNNING" | "SUCCEEDED" | "FAILED";

export interface ReviewQuery {
  product_id: string;
  keyword?: string;
  site?: SiteCode;
  language?: string;
  min_rating?: number;
  max_rating?: number;
  created_from?: string;
  created_to?: string;
  offset?: number;
  limit?: number;
}

export interface ProductReview {
  review_id: string;
  product_id: string;
  site: SiteCode;
  rating: number;
  content: string;
  translated_content: string | null;
  language: string;
  sentiment_hint: string;
  issue_type: string;
  topics?: string[];
  created_at: string;
  source_type: string;
  is_mock_data: boolean;
}

export interface ReviewAnalysisRequest {
  idempotency_key: string;
  product_id: string;
  keyword?: string;
  site?: SiteCode;
  languages: string[];
  min_rating?: number;
  max_rating?: number;
  created_from?: string;
  created_to?: string;
  batch_size: number;
  maximum_reviews: number;
  max_attempts: number;
}

export interface ReviewAnalysisCreated {
  analysis_id: string;
  agent_task_id: string;
  status: AnalysisStatus;
  duplicate: boolean;
  analyzer_version: string;
  analysis_mode: "rule" | "validated_model";
  prompt_version: string | null;
  model_version: string | null;
  is_mock_data: boolean;
}

export interface TopicAggregate {
  topic: string;
  count: number;
  frequency_rate: string;
  negative_count: number;
  severity: string;
}

export interface PainPointAggregate {
  pain_point: string;
  negative_count: number;
  frequency_rate: string;
  severity: string;
  affected_sites: SiteCode[];
}

export interface KeywordAggregate {
  keyword: string;
  count: number;
  review_count: number;
}

export interface TrendPoint {
  site: SiteCode;
  month: string;
  review_count: number;
  negative_count: number;
  average_rating: string;
  topic_counts: Record<string, number>;
}

export interface ReviewJudgement {
  review_id: string;
  product_id: string;
  site: SiteCode;
  rating: number;
  original_content: string;
  translated_content: string | null;
  display_content: string;
  declared_language: string | null;
  detected_language: string;
  translation_status: string;
  sentiment: "positive" | "neutral" | "negative";
  topics: string[];
  confidence: string;
  origin: string;
  source_created_at: string;
  quality_flags: string[];
}

export interface ReviewAnalysisResult extends ReviewAnalysisCreated {
  product_id: string;
  site: SiteCode;
  progress: number;
  current_step: string | null;
  steps: Array<{ step_name: string; status: string; error_message: string | null }>;
  quality: {
    received_count: number;
    included_count: number;
    excluded_count: number;
    flag_counts: Record<string, number>;
    excluded_review_ids: string[];
  } | null;
  sentiment: { positive: number; neutral: number; negative: number } | null;
  topics: TopicAggregate[];
  pain_points: PainPointAggregate[];
  keywords: KeywordAggregate[];
  trends: TrendPoint[];
  judgements: ReviewJudgement[];
  error_message: string | null;
  started_at: string | null;
  finished_at: string | null;
}

export interface ReviewEvidence {
  id: string;
  review_id: string;
  language: string;
  rating: number;
  evidence_type: string;
  label: string;
  sentiment: string | null;
  issue_type: string | null;
  original_content: string;
  translated_content: string | null;
  source_created_at: string;
  confidence: string;
  is_mock_data: boolean;
}

export interface ReviewEvidencePage {
  items: ReviewEvidence[];
  page: number;
  page_size: number;
  total: number;
}

File: backend/src/sellpilot/schemas/review_analysis.py
from datetime import datetime
from decimal import Decimal
from typing import Literal
from uuid import UUID

from pydantic import Field, model_validator

from sellpilot.core.enums import AnalysisStatus, SiteCode, TaskStepStatus
from sellpilot.domain.review_analysis.models import (
    KeywordAggregate,
    PainPointAggregate,
    QualityReport,
    ReviewJudgement,
    SentimentAggregate,
    TopicAggregate,
    TrendPoint,
)
from sellpilot.schemas.common import ContractModel, IdempotencyKey

ReviewAnalysisMode = Literal["rule", "validated_model"]


class ReviewQuery(ContractModel):
    product_id: str = Field(min_length=1, max_length=100)
    site: SiteCode | None = None
    language: str | None = Field(default=None, min_length=2, max_length=32)
    min_rating: int | None = Field(default=None, ge=1, le=5)
    max_rating: int | None = Field(default=None, ge=1, le=5)
    created_from: datetime | None = None
    created_to: datetime | None = None
    offset: int = Field(default=0, ge=0)
    limit: int = Field(default=20, ge=1, le=100)

    @model_validator(mode="after")
    def validate_ranges(self) -> "ReviewQuery":
        if (
            self.min_rating is not None
            and self.max_rating is not None
            and self.min_rating > self.max_rating
        ):
            raise ValueError("min_rating must not exceed max_rating")
        if (
            self.created_from is not None
            and self.created_to is not None
            and self.created_from > self.created_to
        ):
            raise ValueError("created_from must not exceed created_to")
        return self


class ReviewResponse(ContractModel):
    review_id: str
    product_id: str
    site: SiteCode
    rating: int
    content: str
    translated_content: str | None
    language: str
    sentiment_hint: str
    issue_type: str
    created_at: datetime
    source_type: str
    is_mock_data: bool


class ReviewAnalysisCreateRequest(ContractModel):
    idempotency_key: IdempotencyKey
    product_id: str = Field(min_length=1, max_length=100)
    site: SiteCode | None = None
    languages: list[str] = Field(default_factory=list, max_length=10)
    min_rating: int | None = Field(default=None, ge=1, le=5)
    max_rating: int | None = Field(default=None, ge=1, le=5)
    created_from: datetime | None = None
    created_to: datetime | None = None
    batch_size: int = Field(default=100, ge=1, le=100)
    maximum_reviews: int = Field(default=1000, ge=1, le=5000)
    max_attempts: int = Field(default=2, ge=1, le=3)

    @model_validator(mode="after")
    def validate_ranges_and_languages(self) -> "ReviewAnalysisCreateRequest":
        ReviewQuery(
            product_id=self.product_id,
            site=self.site,
            min_rating=self.min_rating,
            max_rating=self.max_rating,
            created_from=self.created_from,
            created_to=self.created_to,
        )
        normalized = [item.strip() for item in self.languages]
        if any(not item or len(item) > 32 for item in normalized):
            raise ValueError("languages must contain non-blank values up to 32 characters")
        if len(normalized) != len(set(normalized)):
            raise ValueError("languages must be unique")
        self.languages[:] = normalized
        return self


class ReviewAnalysisCreatedResponse(ContractModel):
    analysis_id: UUID
    agent_task_id: UUID
    status: AnalysisStatus
    duplicate: bool
    analyzer_version: str
    analysis_mode: ReviewAnalysisMode
    prompt_version: str | None
    model_version: str | None
    is_mock_data: bool


class ReviewTaskStepResponse(ContractModel):
    step_name: str
    status: TaskStepStatus
    input_summary: dict[str, object] | None
    output_summary: dict[str, object] | None
    error_message: str | None
    started_at: datetime | None
    finished_at: datetime | None


class ReviewAnalysisResultResponse(ContractModel):
    analysis_id: UUID
    agent_task_id: UUID | None
    product_id: str
    site: SiteCode
    status: AnalysisStatus
    progress: int = Field(ge=0, le=100)
    current_step: str | None
    steps: tuple[ReviewTaskStepResponse, ...]
    analyzer_version: str
    analysis_mode: ReviewAnalysisMode
    prompt_version: str | None
    model_version: str | None
    quality: QualityReport | None
    sentiment: SentimentAggregate | None
    topics: tuple[TopicAggregate, ...]
    pain_points: tuple[PainPointAggregate, ...]
    keywords: tuple[KeywordAggregate, ...]
    trends: tuple[TrendPoint, ...]
    judgements: tuple[ReviewJudgement, ...]
    error_message: str | None
    is_mock_data: bool
    started_at: datetime | None
    finished_at: datetime | None


class ReviewEvidenceResponse(ContractModel):
    id: UUID
    review_id: str
    product_id: str
    language: str
    rating: int
    evidence_type: str
    label: str
    sentiment: str | None
    issue_type: str | None
    original_content: str
    translated_content: str | None
    source_created_at: datetime
    confidence: Decimal
    metadata: dict[str, object] | None
    is_mock_data: bool


class ReviewEvidencePage(ContractModel):
    items: list[ReviewEvidenceResponse]
    page: int
    page_size: int
    total: int
